src/python/pants/backend/python/providers/python_build_standalone/rules.py

```diff
@@ -0,0 +1,244 @@
+                break
+
+            version_triplet = triplet
+
+    if version_triplet is None:
+        raise Exception(
+            softwrap(
+                f"""\
+                Failed to find a supported Python Build Standalone for Interpreter Constraint: {interpreter_constraints.description}
+
+                Supported versions are currently: {sorted(pbs_versions)}.
+
+                You can teach Pants about newer Python versions supported by Python Build Standalone
+                by setting the `known_python_versions` option in the {PBSPythonProviderSubsystem.name}
+                subsystem. Run `{bin_name()} help-advanced {PBSPythonProviderSubsystem.options_scope}`
+                for more info.
+                """
+            )
+        )
+
+    return ".".join(map(str, version_triplet))
+
+
+@rule
+async def get_python(
+    request: PBSPythonProvider,
+    python_setup: PythonSetup,
+    pbs_subsystem: PBSPythonProviderSubsystem,
+    platform: Platform,
+    named_caches_dir: NamedCachesDirOption,
+    cp: CpBinary,
+) -> PythonExecutable:
+    versions_info = pbs_subsystem.get_all_pbs_pythons()
+
+    python_version = _choose_python(
+        request.interpreter_constraints,
+        python_setup.interpreter_versions_universe,
+        versions_info,
+    )
+    pbs_py_info = versions_info[python_version][platform.value]
+
+    downloaded_python = await Get(
+        DownloadedExternalTool,
+        ExternalToolRequest(
+            DownloadFile(
+                pbs_py_info["url"],
+                FileDigest(
+                    pbs_py_info["sha256"],
+                    pbs_py_info["size"],
+                ),
+            ),
+            "python/bin/python3",
+        ),
+    )
+
+    await Get(
+        ProcessResult,
+        Process(
+            [
+                cp.path,
+                "--recursive",
+                "--no-clobber",
+                "python",
+                f"{PBS_SANDBOX_NAME}/{python_version}",
+            ],
+            level=LogLevel.DEBUG,
+            input_digest=downloaded_python.digest,
+            description=f"Install Python {python_version}",
+            append_only_caches=PBS_APPEND_ONLY_CACHES,
+            # Don't cache, we want this to always be run so that we can assume for the rest of the
+            # session the named_cache destination for this Python is valid, as the Python ecosystem
+            # mainly assumes absolute paths for Python interpreters.
+            cache_scope=ProcessCacheScope.PER_SESSION,
+        ),
+    )
+
+    python_path = named_caches_dir.val / PBS_NAMED_CACHE_NAME / python_version / "bin" / "python3"
+    return PythonExecutable(
+        path=str(python_path),
+        fingerprint=None,
+        # One would normally set append_only_caches=PBS_APPEND_ONLY_CACHES
+        # here, but it is already going to be injected into the pex
+        # environment by PythonBuildStandaloneBinary
```

Comment:
This was to avoid this FrozenDict exploding way down the line.  I'm pretty sure this comment is truthful, but there was much head scratching along the way.

https://github.com/pantsbuild/pants/blob/78f543a9d51ee30396e2d154464a3bafec191566/src/python/pants/backend/python/util_rules/pex.py#L1284